### PR TITLE
Allow specifying xcframeworks in calls to apple_library

### PR DESCRIPTION
**Pull request by segiddins**
Used by https://github.com/bazel-ios/cocoapods-bazel/pull/32

### CODE REVIEW

File: rules/library.bzl [hunk 1]
```diff
@@ -219,6 +220,82 @@ def _canonicalize_swift_version(swift_version):
...
+        ))
+    return (platform, platform_variant, supported_archs, import_name)
+
+def _xcframework(*, library_name, name, slices):
+    xcframework_name = "{}-import-{}.xcframework".format(library_name, name)
+    conditions = {}
+    for slice in slices:
+        platform, platform_variant, archs, name = _xcframework_slice(xcframework_name = xcframework_name, **slice)
+        platform_setting = "@build_bazel_rules_ios//rules/apple_platform:" + platform
+        # platform_variant_setting = "@build_bazel_rules_ios//rules/apple_platform:platform_variant_" + platform_variant if platform_variant else None
```
Comment (congt) [inline rules/library.bzl:263]: nit: remove

**Review by congt — APPROVED** [review]

**Review by thiagohmcruz — APPROVED** [review]

File: rules/library.bzl [hunk 2]
```diff
@@ -219,6 +220,82 @@ def _canonicalize_swift_version(swift_version):
...
+
+        for arch in archs:
+            if arch == "arm64" and platform_variant == "simulator":
+                # TODO: support sim on apple silicon by having a config setting for platform_variant
+                continue
+
+            arch_setting = "@build_bazel_rules_apple//apple:{}_{}".format(platform, arch)
```
Comment (jerrymarino) [inline rules/library.bzl:270]: Any reason to not put this in a select?

File: rules/library.bzl [hunk 3]
```diff
@@ -219,6 +220,82 @@ def _canonicalize_swift_version(swift_version):
...
+    else:
+        fail("Unsupported xcframework slice type {} ({}) in {}".format(
+            build_type,
+            identifier,
+            xcframework_name,
+        ))
+    return (platform, platform_variant, supported_archs, import_name)
+
+def _xcframework(*, library_name, name, slices):
+    xcframework_name = "{}-import-{}.xcframework".format(library_name, name)
```
Comment (jerrymarino) [inline rules/library.bzl:258]: So we add a dep on xcframework per slice? Do apple not have a notion of fat multi platform xcframeworks - or is this uncommon?

File: rules/library.bzl [hunk 4]
```diff
@@ -219,6 +220,82 @@ def _canonicalize_swift_version(swift_version):
...
     return swift_version
 
+def _xcframework_build_type(*, linkage, packaging):
+    return (linkage, packaging)
+
+def _xcframework_slice(*, xcframework_name, identifier, platform, platform_variant, supported_archs, build_type, path):
+    linkage, packaging = _xcframework_build_type(**build_type)
+    import_name = "{}-{}".format(xcframework_name, identifier)
+    if (linkage, packaging) == ("dynamic", "framework"):
+        apple_dynamic_framework_import(
```
Comment (jerrymarino) [inline rules/library.bzl:230]: See lower comment, isn't this a multi-arch or mutli platform handling rule?

**Review by jerrymarino — APPROVED** [review]
LGTM - left some questions to ponder regarding fat and multi arch

**Comment by jerrymarino** [comment]
Also any thoughts on adding this into `rules_apple` ?

File: rules/library.bzl (hunk 2 above)
Comment (segiddins) [reply rules/library.bzl:270]: it's used in the condition of the select, created by the `selects` library in skylib

File: rules/library.bzl (hunk 3 above)
Comment (segiddins) [reply rules/library.bzl:258]: they can be fat -- they can support multiple archs

**Comment by segiddins** [comment]
rules_apple tries to focus on providing rules, rather than macros

File: rules/library.bzl (hunk 4 above)
Comment (segiddins) [reply rules/library.bzl:230]: it is multi-arch, and indeed we're allowing it to handle slices that have multiple archs

**Pull request merged**